### ISSUE TITLE
Snaps failing to build  on build.snapcraft.io

### ISSUE THREAD

**Issue by frederickjh** [frederickjh/qxmledit#17]
After an email from the Snap Store that QXmlEdit contains outdated Ubuntu packages, the builds on all architectures are failing to build on build .snapcraft.io. Here is the last part of the log for the amd64 build that failed:

```

g++ -c -include ../build/lib/obj/QXmlEditWidget-0.9.14 -m64 -pipe -Wall -Wno-deprecated-declarations -Werror=format-security -Werror -O2 -Wall -W -D_REENTRANT -fPIC -DLIBQXMLEDIT_LIBRARY -DUNIX_RESOURCES -DUNIX_RESOURCE_PATH=/snap/qxmledit/current/share/qxmledit -DUNIX_DOC_PATH=/snap/qxmledit/current/share/qxmledit/doc -DQT_NO_DEBUG -DQT_SVG_LIB -DQT_PRINTSUPPORT_LIB -DQT_WIDGETS_LIB -DQT_GUI_LIB -DQT_XML_LIB -DQT_XMLPATTERNS_LIB -DQT_QML_LIB -DQT_NETWORK_LIB -DQT_CORE_LIB -I. -I../../install/usr/include -I. -Iglobals/includes -I./globals/framework -Iglobals/includes -Iframework/include -isystem /usr/include/x86_64-linux-gnu/qt5 -isystem /usr/include/x86_64-linux-gnu/qt5/QtSvg -isystem /usr/include/x86_64-linux-gnu/qt5/QtPrintSupport -isystem /usr/include/x86_64-linux-gnu/qt5/QtWidgets -isystem /usr/include/x86_64-linux-gnu/qt5/QtGui -isystem /usr/include/x86_64-linux-gnu/qt5/QtXml -isystem /usr/include/x86_64-linux-gnu/qt5/QtXmlPatterns -isystem /usr/include/x86_64-linux-gnu/qt5/QtQml -isystem /usr/include/x86_64-linux-gnu/qt5/QtNetwork -isystem /usr/include/x86_64-linux-gnu/qt5/QtCore -I../build/lib/moc -I../build/lib/ui -I/usr/lib/x86_64-linux-gnu/qt5/mkspecs/linux-g++-64 -o ../build/lib/obj/extractionscriptmanager.o extraction/scripting/extractionscriptmanager.cpp
extraction/scripting/extractionscriptmanager.cpp:24:33: fatal error: QtScript/QScriptValue: No such file or directory
compilation terminated.
make[1]: *** [../build/lib/obj/extractionscriptmanager.o] Error 1
make[1]: *** Waiting for unfinished jobs....
Makefile.QXmlEditWidget:11751: recipe for target '../build/lib/obj/extractionscriptmanager.o' failed
make[1]: Leaving directory '/build/qxmledit/parts/qxmledit/build/src'
Makefile:88: recipe for target 'sub-src-QXmlEditWidget-pro-make_first-ordered' failed
make: *** [sub-src-QXmlEditWidget-pro-make_first-ordered] Error 2
qmake LIBS+="-L/build/qxmledit/parts/qxmledit/install/lib" LIBS+="-L/build/qxmledit/parts/qxmledit/install/usr/lib" LIBS+="-L/build/qxmledit/parts/qxmledit/install/lib/x86_64-linux-gnu" LIBS+="-L/build/qxmledit/parts/qxmledit/install/usr/lib/x86_64-linux-gnu" INCLUDEPATH+="/build/qxmledit/parts/qxmledit/install/usr/include" CONFIG+=release QXMLEDIT_INST_DIR=/snap/qxmledit/current/bin QXMLEDIT_INST_DATA_DIR=/snap/qxmledit/current/share/qxmledit QXMLEDIT_INST_DOC_DIR=/snap/qxmledit/current/share/qxmledit/doc QXMLEDIT_INST_LIB_DIR=/snap/qxmledit/current/lib QXMLEDIT_INST_INCLUDE_DIR=/snap/qxmledit/current/share/qxmledit/include
make -j4
Failed to run 'make -j4' for 'qxmledit': Exited with code 2.
Verify that the part is using the correct parameters and try again.
sed: can't read /build/qxmledit/parts/qxmledit/install/snap/qxmledit/current/share/qxmledit/QXmlEdit.desktop: No such file or directory
Build failed
Traceback (most recent call last):
  File "/usr/lib/python2.7/dist-packages/lpbuildd/target/build_snap.py", line 266, in run
    self.build()
  File "/usr/lib/python2.7/dist-packages/lpbuildd/target/build_snap.py", line 255, in build
    env=env)
  File "/usr/lib/python2.7/dist-packages/lpbuildd/target/build_snap.py", line 102, in run_build_command
    return self.backend.run(args, env=full_env, **kwargs)
  File "/usr/lib/python2.7/dist-packages/lpbuildd/target/lxd.py", line 527, in run
    subprocess.check_call(cmd, **kwargs)
  File "/usr/lib/python2.7/subprocess.py", line 541, in check_call
    raise CalledProcessError(retcode, cmd)
CalledProcessError: Command '['lxc', 'exec', 'lp-xenial-amd64', '--env', 'LANG=C.UTF-8', '--env', 'SHELL=/bin/sh', '--env', 'SNAPCRAFT_BUILD_INFO=1', '--env', 'SNAPCRAFT_IMAGE_INFO={"build-request-id": "lp-53737680", "build-request-timestamp": "2020-01-10T20:44:39Z", "build_url": "https://launchpad.net/~build.snapcraft.io/+snap/3c0bf9352d8f6a70a84ca152eac46724/+build/791485"}', '--env', 'SNAPCRAFT_BUILD_ENVIRONMENT=host', '--env', 'http_proxy=http://10.10.10.1:8222/', '--env', 'https_proxy=http://10.10.10.1:8222/', '--env', 'GIT_PROXY_COMMAND=/usr/local/bin/snap-git-proxy', '--', '/bin/sh', '-c', 'cd /build/qxmledit && linux64 snapcraft']' returned non-zero exit status 2
Revoking proxy token...
RUN: /usr/share/launchpad-buildd/bin/in-target scan-for-processes --backend=lxd --series=xenial --arch=amd64 SNAPBUILD-791485
Scanning for processes to kill in build SNAPBUILD-791485
```

**Comment by lbellonda** [comment]
Hello, 
      this issue can be related to a deprecated module in Qt, QtScript.
It can be related to this issue: https://github.com/lbellonda/qxmledit/issues/66
Can you test the "devel" branch?
Thank you.

**Comment by frederickjh** [comment]
@lbellonda I just tried that. I am once again having the issue with the same files being included in both the `qxmledit` and `desktop-qt5` packages.

Here is the diff of the change I tried. I switched to the **devel** branch and then removed one fileset exclude for the locales as it seems to no longer be needed and added several rules to exclude the files that are double.

```
$ git diff
diff --git a/snap/snapcraft.yaml b/snap/snapcraft.yaml
index d1757f7..4b42aeb 100644
--- a/snap/snapcraft.yaml
+++ b/snap/snapcraft.yaml
@@ -45,6 +45,7 @@ parts:
       - QXMLEDIT_INST_LIB_DIR=/snap/$SNAPCRAFT_PROJECT_NAME/current/lib
       - QXMLEDIT_INST_INCLUDE_DIR=/snap/$SNAPCRAFT_PROJECT_NAME/current/share/$SNAPCRAFT_PROJECT_NAME/include
     source: https://github.com/lbellonda/qxmledit.git
+    source-branch: devel
     source-depth: 50
     build-packages:
       - build-essential
@@ -76,8 +77,13 @@ parts:
       - libstdc++6
       - zlib1g
     filesets:
-      exclude-x11-locales:
-      - -usr/share/x11/locale/*
+      exclude-qt5-plugins:
+      - -usr/lib/x86_64-linux-gnu/qt5/plugins/*/*.so
+      exclude-x86-64-linux-gnu-usr-libs:
+      - -usr/lib/x86_64-linux-gnu/glib*
+      - -usr/lib/x86_64-linux-gnu/lib*
+      exclude-x86-64-linux-gnu-libs:
+      - -lib/x86_64-linux-gnu/lib*
       exclude-changelogs:
       - -usr/share/doc/*/changelog*
     organize:
```

Here is the error after running `snapcraft` to build the snap locally:

```
Failed to stage: Parts 'desktop-qt5' and 'qxmledit' have the following files, but with different contents:
    lib/x86_64-linux-gnu/libdbus-1.so.3.14.6
    lib/x86_64-linux-gnu/libexpat.so.1.6.0
    lib/x86_64-linux-gnu/libglib-2.0.so.0.4800.2
    usr/lib/x86_64-linux-gnu/glib-2.0/gio-querymodules
    usr/lib/x86_64-linux-gnu/glib-2.0/glib-compile-resources
    usr/lib/x86_64-linux-gnu/glib-2.0/glib-compile-schemas
    usr/lib/x86_64-linux-gnu/libQt5Core.so.5.5.1
    usr/lib/x86_64-linux-gnu/libQt5DBus.so.5.5.1
    usr/lib/x86_64-linux-gnu/libQt5EglDeviceIntegration.so.5.5.1
    usr/lib/x86_64-linux-gnu/libQt5Gui.so.5.5.1
    usr/lib/x86_64-linux-gnu/libQt5Network.so.5.5.1
    usr/lib/x86_64-linux-gnu/libQt5Widgets.so.5.5.1
    usr/lib/x86_64-linux-gnu/libQt5XcbQpa.so.5.5.1
    usr/lib/x86_64-linux-gnu/libavahi-client.so.3.2.9
    usr/lib/x86_64-linux-gnu/libavahi-common.so.3.5.3
    usr/lib/x86_64-linux-gnu/libcups.so.2
    usr/lib/x86_64-linux-gnu/libelf-0.165.so
    usr/lib/x86_64-linux-gnu/libexpatw.so.1.6.0
    usr/lib/x86_64-linux-gnu/libfreetype.so.6.12.1
    usr/lib/x86_64-linux-gnu/libgio-2.0.so.0.4800.2
    usr/lib/x86_64-linux-gnu/libgmodule-2.0.so.0.4800.2
    usr/lib/x86_64-linux-gnu/libgnutls.so.30.6.2
    usr/lib/x86_64-linux-gnu/libgobject-2.0.so.0.4800.2
    usr/lib/x86_64-linux-gnu/libgssapi_krb5.so.2.2
    usr/lib/x86_64-linux-gnu/libgthread-2.0.so.0.4800.2
    usr/lib/x86_64-linux-gnu/libjpeg.so.8.0.2
    usr/lib/x86_64-linux-gnu/libk5crypto.so.3.1
    usr/lib/x86_64-linux-gnu/libkrb5.so.3.3
    usr/lib/x86_64-linux-gnu/libkrb5support.so.0.1
    usr/lib/x86_64-linux-gnu/libsqlite3.so.0.8.6
    usr/lib/x86_64-linux-gnu/qt5/plugins/bearer/libqconnmanbearer.so
    usr/lib/x86_64-linux-gnu/qt5/plugins/bearer/libqgenericbearer.so
    usr/lib/x86_64-linux-gnu/qt5/plugins/bearer/libqnmbearer.so
    usr/lib/x86_64-linux-gnu/qt5/plugins/egldeviceintegrations/libqeglfs-kms-integration.so
    usr/lib/x86_64-linux-gnu/qt5/plugins/egldeviceintegrations/libqeglfs-x11-integration.so
    usr/lib/x86_64-linux-gnu/qt5/plugins/generic/libqevdevkeyboardplugin.so
    usr/lib/x86_64-linux-gnu/qt5/plugins/generic/libqevdevmouseplugin.so
    usr/lib/x86_64-linux-gnu/qt5/plugins/generic/libqevdevtabletplugin.so
    usr/lib/x86_64-linux-gnu/qt5/plugins/generic/libqevdevtouchplugin.so
    usr/lib/x86_64-linux-gnu/qt5/plugins/generic/libqlibinputplugin.so
    usr/lib/x86_64-linux-gnu/qt5/plugins/generic/libqtuiotouchplugin.so
    usr/lib/x86_64-linux-gnu/qt5/plugins/imageformats/libqgif.so
    usr/lib/x86_64-linux-gnu/qt5/plugins/imageformats/libqico.so
    usr/lib/x86_64-linux-gnu/qt5/plugins/imageformats/libqjpeg.so
    usr/lib/x86_64-linux-gnu/qt5/plugins/platforminputcontexts/libcomposeplatforminputcontextplugin.so
    usr/lib/x86_64-linux-gnu/qt5/plugins/platforminputcontexts/libibusplatforminputcontextplugin.so
    usr/lib/x86_64-linux-gnu/qt5/plugins/platforms/libqeglfs.so
    usr/lib/x86_64-linux-gnu/qt5/plugins/platforms/libqlinuxfb.so
    usr/lib/x86_64-linux-gnu/qt5/plugins/platforms/libqminimal.so
    usr/lib/x86_64-linux-gnu/qt5/plugins/platforms/libqminimalegl.so
    usr/lib/x86_64-linux-gnu/qt5/plugins/platforms/libqoffscreen.so
    usr/lib/x86_64-linux-gnu/qt5/plugins/platforms/libqxcb.so
    usr/lib/x86_64-linux-gnu/qt5/plugins/xcbglintegrations/libqxcb-egl-integration.so
    usr/lib/x86_64-linux-gnu/qt5/plugins/xcbglintegrations/libqxcb-glx-integration.so
    usr/share/doc/libavahi-common-data/changelog.Debian.gz
    usr/share/doc/libcups2/changelog.Debian.gz
    usr/share/doc/libdbus-1-3/changelog.Debian.gz
    usr/share/doc/libelf1/changelog.Debian.gz
    usr/share/doc/libexpat1/changelog.Debian.gz
    usr/share/doc/libfreetype6/changelog.Debian.gz
    usr/share/doc/libglib2.0-0/changelog.Debian.gz
    usr/share/doc/libgnutls30/changelog.Debian.gz
    usr/share/doc/libjpeg-turbo8/changelog.Debian.gz
    usr/share/doc/libkrb5support0/changelog.Debian.gz
    usr/share/doc/libqt5core5a/changelog.Debian.gz
    usr/share/doc/libsqlite3-0/changelog.Debian.gz

Snapcraft offers some capabilities to solve this by use of the following keywords:
    - `filesets`
    - `stage`
    - `snap`
    - `organize`

To learn more about these part keywords, run `snapcraft help plugins`.
```

I am not sure why the excludes are not working, including the one that was working for the changelog files. Any help on this from anyone would be appreciated.

**Comment by frederickjh** [comment]
@Lin-Buo-Ren I know you helped me last time with the fileset excludes and not needing to specify every filename individually. Maybe you could take a quick look at what I tried and maybe give me a nudge in the right direction to solve this.

**Comment by frederickjh** [comment]
I changed the title as since the original email saying that `libgnutls30` package in the snaps  is out-dated I have received emails for the `libbsd0`, `zlib1g`, and another email for a new update to `libgnutls30`.

I believe that @lbellonda is correct in that it could be related to a deprecated module in Qt, QtScript.
However, I have not been able to solve the

> Failed to stage: Parts 'desktop-qt5' and 'qxmledit'have the following files, but with different contents

issue, as the filesets excludes do not seem to work as expected and the documentation on this is very vague at best, or at least I have not found any that seems to explain it in a way I can understand.

**Comment by frederickjh** [comment]
@lbellonda  I think I managed to get this working. Yes it is related to https://github.com/lbellonda/qxmledit/issues/66. The `devel` branch snaps but the `master` branch does not.

I will release QXmlEdit version 0.9.14 in the **edge** snap channel. If you can test it out that would be appreciated, as I no longer have a reason to use QXmlEdit in my work. I will let you know when the builds finish and I have released to the **edge** snap channel for QXmlEdit.

Once you release the next version of QXmlEdit I can switch the snap to use the `master` branch and release to **stable** snap channel.

Although Github closed this issue because I merge to master the snap is in **development** mode. The grade was switched from **master** to **devel** as we switched from using the `master` branch to using the `devel` branch.

**Comment by frederickjh** [comment]
@lbellonda The snap is up and available in the Snap Store on the **edge** channel. You can install it with:

`sudo snap install qxmledit --edge`

If you have already installed the QXmlEdit snap from the **stable** channel you can switch to the version in the **edge** channel (**0.9.14-4-gc41384a**) with:

`sudo snap refresh --channel=edge qxmledit`

to switch back to stable (**0.19.13**):

`sudo snap refresh --channel=stable qxmledit`

**Comment by frederickjh** [comment]
I got an email that `libmysqlclient20` and `mysql-common` packages had security updates available. I rebuilt the snap and have released it into the dev channel.

Once the next stable release of QXmlEdit gets released (0.9.15 ?) I will switch this snap back to the master branch of [lbellonda/qxmledit](https://github.com/lbellonda/qxmledit) and then release on the [QXmlEdit stable channel in the Snap Store](https://snapcraft.io/qxmledit).
